controllers/tortoise_controller_test.go

```diff
@@ -291,7 +310,222 @@ var _ = Describe("Test TortoiseController", func() {
 			}).Should(Succeed())
 		})
 		It("TortoisePhaseEmergency", func() {
-			// TODO: add
+			now := time.Now()
+			tc := testCase{
+				before: resources{
+					tortoise: utils.NewTortoiseBuilder().
+						SetName("mercari").
+						SetNamespace("default").
+						SetTargetRefs(v1alpha1.TargetRefs{
+							DeploymentName:              "mercari-app",
+							HorizontalPodAutoscalerName: pointer.String("hpa"),
+						}).
+						SetUpdateMode(v1alpha1.UpdateModeEmergency).
+						AddResourcePolicy(v1alpha1.ContainerResourcePolicy{
+							ContainerName: "app",
+							AutoscalingPolicy: map[corev1.ResourceName]v1alpha1.AutoscalingType{
+								corev1.ResourceCPU:    v1alpha1.AutoscalingTypeHorizontal,
+								corev1.ResourceMemory: v1alpha1.AutoscalingTypeVertical,
+							},
+						}).
+						SetTortoisePhase(v1alpha1.TortoisePhaseWorking). // will be updated.
+						SetRecommendations(v1alpha1.Recommendations{
+							Horizontal: &v1alpha1.HorizontalRecommendations{
+								TargetUtilizations: []v1alpha1.HPATargetUtilizationRecommendationPerContainer{
+									{
+										ContainerName: "app",
+										TargetUtilization: map[corev1.ResourceName]int32{
+											corev1.ResourceCPU: 50, // will be updated.
+										},
+									},
+								},
+								MaxReplicas: []v1alpha1.ReplicasRecommendation{
+									{
+										From:      0,
+										To:        24,
+										WeekDay:   now.Weekday().String(),
+										TimeZone:  now.Location().String(),
+										Value:     15, // will be updated
+										UpdatedAt: metav1.NewTime(now),
+									},
+								},
+								MinReplicas: []v1alpha1.ReplicasRecommendation{
+									{
+										From:      0,
+										To:        24,
+										WeekDay:   now.Weekday().String(),
+										TimeZone:  now.Location().String(),
+										Value:     3, // will be updated
+										UpdatedAt: metav1.NewTime(now),
+									},
+								},
+							},
+						}).
+						AddCondition(v1alpha1.ContainerRecommendationFromVPA{
+							ContainerName: "app",
+							Recommendation: map[corev1.ResourceName]v1alpha1.ResourceQuantity{
+								corev1.ResourceCPU:    {},
+								corev1.ResourceMemory: {},
+							},
+							MaxRecommendation: map[corev1.ResourceName]v1alpha1.ResourceQuantity{
+								corev1.ResourceCPU:    {},
+								corev1.ResourceMemory: {},
+							},
+						}).
+						Build(),
+					deployment: deploymentWithReplicaNum(10),
+				},
+			}
+
+			err := tc.createResources(ctx, k8sClient, cfg)
+			Expect(err).ShouldNot(HaveOccurred())
+
+			// create the desired HPA from the created definition.
+			wantHPA := tc.before.hpa.DeepCopy()
+			wantHPA.Spec.MinReplicas = pointer.Int32(20)
+			wantHPA.Spec.MaxReplicas = 20
+			for i, m := range wantHPA.Spec.Metrics {
+				if m.External != nil && m.External.Metric.Name == "datadogmetric@default:mercari-app-cpu-app" {
+					wantHPA.Spec.Metrics[i].External.Target.Value = resourceQuantityPtr(resource.MustParse("90"))
```

Comment:
@sanposhiho Thanks, done! BTW just curious why is it a bug? I think it won't make any difference if the targetUtilization is 0% or 100% when minReplica == maxReplica

Reply:
That was this integration test specific bug. 

> it won't make any difference if the targetUtilization is 0% or 100% when minReplica == maxReplica

This is true. But, tortoise's expected behavior here is updating the target utilization like the tortoise does during the non-emergency mode although, in actual case, any target utilization doesn't give any impact on HPA's behavior. 

---

To exactly explain how this bug happened (sorry super long story):

We [create](https://github.com/mercari/tortoise/blob/f4f350cc105214c5474bd31af5d4d95677de5888/controllers/tortoise_controller_test.go#L398) tortoise and then [update](https://github.com/mercari/tortoise/blob/f4f350cc105214c5474bd31af5d4d95677de5888/controllers/tortoise_controller_test.go#L408) tortoise status later in `createResources` fn in this integration test.
So, after creating tortoise and before updating status, the controller will try to reconcile the tortoise. 
This test case specifies the emergency mode, so the tortoise controller regards the tortoise as finishing the [initialization](https://github.com/mercari/tortoise/blob/main/pkg/tortoise/tortoise.go#L98) step and tries to proceed creating the recommendation on the tortoise status. (In initialization step, the controller fills some required fields in tortoise's status)

But, the initialization step isn't actually finished and updating the tortoise status get failed because some required field in status are missing. Before introducing [this bug fix](https://github.com/mercari/tortoise/pull/60), the controller is handling the tortoise like:
1. calculate the recommendation
2. update HPA based on calculated recommendation. 
3. update tortoise status with calculated recommendations.


So, if failing to update tortoise status, HPA and VPA have been updated based on calculated recommendations, but tortoise status cannot get the recommendation on the status. 

Also, during (2), In this test case, the CPU target utilization gets changed 50 -> 75 because: 
- the VPA recommendation is 3 core, the current HPA target utilization is 50% (=2 core), and the CPU request is 4 cores. 
- VPA recommendation showed that the target Pod is sometimes using 1 more core than the target utilization. In other words, we only need to give 1 additional core (= 25% given the CPU req is 4 core)

Then after updating tortoise status, the controller again reconciles the tortoise; calculate the recommendation again. The tortoise is calculating the target utilization recommendation based on the VPA's recommendation value and HPA's current target utilization.
But, the HPA was updated at the last reconciliation (50% -> 75%), but the VPA doesn't change its recommendation because the VPA in the integration test is just fake one.
So the recommendation is calculated like:
- the VPA recommendation is 3 core (not updated), the current HPA target utilization is 75% (=3 core), and the CPU request is 4 cores. 
- VPA recommendation showed that the target Pod is using exactly the same core as the HPA target utilization. In other words, we don't need to give additional core to this Pod. (shouldn't actually happen)

So, the tortoise recommend 100% CPU target utilization on HPA in this reconciliation, but we configure that 90% is the maximum, so at final, the tortoise gives 90% target utilization on HPA.
 

[This bug fix](https://github.com/mercari/tortoise/pull/60) changed the controller to update the status once before updating HPA and VPA, so we can now prevent this from happening.